### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries/DirichletContinuation): results on logarithmic derivatives

### PR DESCRIPTION
**Pull request by MichaelStollBayreuth**
As a further step in the direction of Dirichlet's Theorem on primes in AP, this PR adds results on (negative) logarithmic derivatives of `L χ` for a non-trivial Dirichlet character `χ` and of `s ↦ (L χ s) * (s - 1)` when `χ` is trivial.

See [here on Zulip](https://leanprover.zulipchat.com/#narrow/channel/144837-PR-reviews/topic/Prerequisites.20for.20PNT.20and.20Dirichlet's.20Thm/near/482543603).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 1]
```diff
@@ -307,3 +309,91 @@ theorem completedLFunction_one_sub {χ : DirichletCharacter ℂ N} (hχ : IsPrim
...
+  simp only [ne_eq, hs, not_false_eq_true, Function.update_noteq]
+
+lemma LFunctionTrivChar₁_apply_one_ne_zero : LFunctionTrivChar₁ n 1 ≠ 0 := by
+  simp only [Function.update_same]
+  refine Finset.prod_ne_zero_iff.mpr fun p hp ↦ ?_
+  rw [sub_ne_zero, ne_eq, one_eq_inv]
+  exact_mod_cast (Nat.prime_of_mem_primeFactors hp).ne_one
+
+/-- `s ↦ (s - 1) * L χ s` is an entire function when `χ` is a trivial Dirichlet character. -/
+lemma LFunctionTrivChar₁_differentiable : Differentiable ℂ (LFunctionTrivChar₁ n) := by
```
Comment (loefflerd) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:347]: Maybe `differentiable_LFunctionTrivChar1`, for consistency with `deriv_LFunction[xxx]` below?

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 2]
```diff
@@ -307,3 +309,91 @@ theorem completedLFunction_one_sub {χ : DirichletCharacter ℂ N} (hχ : IsPrim
...
+  rw [← differentiableOn_univ,
+    ← differentiableOn_compl_singleton_and_continuousAt_iff (c := 1) Filter.univ_mem]
+  refine ⟨DifferentiableOn.congr (f := fun s ↦ (s - 1) * LFunctionTrivChar n s)
+    (fun _ hs ↦ DifferentiableAt.differentiableWithinAt <| by fun_prop (disch := simp_all [hs]))
+    fun _ hs ↦ by rw [LFunctionTrivChar₁_apply_of_ne_one n (Set.mem_diff_singleton.mp hs).2],
```
Comment (loefflerd) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:352]:
```suggestion
    fun _ hs ↦ Function.update_noteq (Set.mem_diff_singleton.mp hs).2 ..
```

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 3]
```diff
@@ -307,3 +309,91 @@ theorem completedLFunction_one_sub {χ : DirichletCharacter ℂ N} (hχ : IsPrim
...
+    using LFunctionTrivChar_residue_one
+
+lemma deriv_LFunctionTrivChar₁_apply_of_ne_one {s : ℂ} (hs : s ≠ 1) :
+    deriv (LFunctionTrivChar₁ n) s =
+      (s - 1) * deriv (LFunctionTrivChar n) s + LFunctionTrivChar n s := by
+  have H : deriv (LFunctionTrivChar₁ n) s =
+      deriv (fun w ↦ (w - 1) * LFunctionTrivChar n w) s := by
+    refine Filter.eventuallyEq_iff_exists_mem.mpr ?_ |>.deriv_eq
+    exact ⟨{w | w ≠ 1}, IsOpen.mem_nhds isOpen_ne hs,
+      fun w hw ↦ LFunctionTrivChar₁_apply_of_ne_one n (Set.mem_setOf.mp hw)⟩
```
Comment (loefflerd) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:364]:
```suggestion
    refine eventuallyEq_iff_exists_mem.mpr ?_ |>.deriv_eq
    exact ⟨_, isOpen_ne.mem_nhds hs, fun _ hw ↦ Function.update_noteq (Set.mem_setOf.mp hw) ..⟩
```

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 4]
```diff
@@ -307,3 +309,91 @@ theorem completedLFunction_one_sub {χ : DirichletCharacter ℂ N} (hχ : IsPrim
...
+lemma continuousOn_neg_logDeriv_LFunctionTrivChar₁ :
+    ContinuousOn (fun s ↦ -deriv (LFunctionTrivChar₁ n) s / LFunctionTrivChar₁ n s)
+      {s | s = 1 ∨ LFunctionTrivChar n s ≠ 0} := by
+  simp_rw [neg_div]
+  have h := LFunctionTrivChar₁_differentiable n
+  refine ((h.contDiff.continuous_deriv le_rfl).continuousOn.div
+    h.continuous.continuousOn fun w hw ↦ ?_).neg
+  rcases eq_or_ne w 1 with rfl | hw'
+  · exact LFunctionTrivChar₁_apply_one_ne_zero _
+  · rw [LFunctionTrivChar₁_apply_of_ne_one n hw', mul_ne_zero_iff]
```
Comment (loefflerd) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:379]:
```suggestion
  · rw [LFunctionTrivChar₁, Function.update_noteq hw', mul_ne_zero_iff]
```

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 5]
```diff
@@ -307,3 +309,91 @@ theorem completedLFunction_one_sub {χ : DirichletCharacter ℂ N} (hχ : IsPrim
...
+
+section nontrivial
+
+variable {n : ℕ} [NeZero n] {χ : DirichletCharacter ℂ n}
+
+/-- The negative logarithmic derivative of the L-function of a nontrivial Dirichlet character
+is continuous away from the zeros of the L-function. -/
+lemma continuousOn_neg_logDeriv_LFunction_of_nontriv (hχ : χ ≠ 1) :
+    ContinuousOn (fun s ↦ -deriv (LFunction χ) s / LFunction χ s) {s | LFunction χ s ≠ 0} := by
+  simp_rw [neg_div]
```
Comment (loefflerd) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:392]:
```suggestion
  simp only [neg_div]
```
(saves a few hundred heartbeats)

**Review by loefflerd — COMMENTED** [review]
This looks great! I found a way of optimising out one lemma, and a few minor golfs. I'm really happy that we're closing in on PNT and PNT-AP now.

maintainer delegate

**Review by jcommelin — COMMENTED** [review]
bors d+

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean [hunk 6]
```diff
@@ -307,3 +309,91 @@ theorem completedLFunction_one_sub {χ : DirichletCharacter ℂ N} (hχ : IsPrim
...
+/-- The function obtained by "multiplying away" the pole of `L χ` for a trivial Dirichlet
+character `χ`. Its (negative) logarithmic derivative is used to prove Dirichlet's Theorem
+on primes in arithmetic progression. -/
+noncomputable abbrev LFunctionTrivChar₁ : ℂ → ℂ :=
+  Function.update (fun s ↦ (s - 1) * LFunctionTrivChar n s) 1
+    (∏ p ∈ n.primeFactors, (1 - (p : ℂ)⁻¹))
+
+lemma LFunctionTrivChar₁_apply_of_ne_one {s : ℂ} (hs : s ≠ 1) :
+    LFunctionTrivChar₁ n s = (s - 1) * LFunctionTrivChar n s := by
+  simp only [ne_eq, hs, not_false_eq_true, Function.update_noteq]
```
[the remark this answers is not in the document]
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:338]: Inlining doesn't work so well, so I'll keep the lemma with the simpler proof.

[the remark this answers is not in the document]
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:338]: OK, it does, asa you've demonstrated :smile: . So I'm removing it, after all...

**Comment by MichaelStollBayreuth** [comment]
@loefflerd Thanks! I made the suggested changes.

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean (hunk 6 above)
[the remark this answers is not in the document]
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/DirichletContinuation.lean]: What is it that doesn’t work? I left suggestions on the PR detailing how to inline it at each of the places the lemma was used.

[the remark this answers is not in the document]
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/DirichletContinuation.lean]: (I was just being lazy; replacing by the inlined proof gave errors...)

**Pull request closed without merging**